layers.go

```diff
@@ -823,33 +823,45 @@ func (r *layerStore) Diff(from, to string, options *DiffOptions) (io.ReadCloser,
 	if options != nil && options.Compression != nil {
 		compression = *options.Compression
 	}
+	maybeCompressReadCloser := func(rc io.ReadCloser) (io.ReadCloser, error) {
+		if compression == archive.Uncompressed {
+			return rc, nil
+		}
+		preader, pwriter := io.Pipe()
+		compressor, err := archive.CompressStream(pwriter, compression)
+		if err != nil {
+			rc.Close()
+			pwriter.Close()
+			preader.Close()
+			return nil, err
+		}
+		go func() {
+			io.Copy(compressor, rc)
+			rc.Close()
+			compressor.Close()
```

Comment:
Maybe?  Are we worried about cleaning up if Copy() panics?  If not, then deferring them inside the anonymous goroutine doesn't buy us anything.